providers/implementations/kdfs/argon2.c

```diff
@@ -0,0 +1,1488 @@
+}
+
+static int validate_inputs(const ARGON2_CTX *ctx)
+{
+    /*
+     * due to -Werror=type-limits, some of the comparisons are made as
+     * X+1 > Y+1 rather than X > Y, as this caused problems when some
+     * lower limits were set to zero.
+     */
+
+    if (ctx == NULL)
+        return 0;
+
+    switch (ctx->type) {
+    case ARGON2_D:
+    case ARGON2_I:
+    case ARGON2_ID:
+        break;
+    default:
+        return 0;
+    }
+
+    if (ctx->out == NULL)
+        return 0;
+    if (ctx->outlen < ARGON2_MIN_OUTLEN)
+        return 0;
+    if (ctx->outlen > ARGON2_MAX_OUTLEN)
+        return 0;
+    if (ctx->pwd == NULL && ctx->pwdlen != 0)
+        return 0;
+    if (ctx->pwdlen+1 < ARGON2_MIN_PWD_LENGTH+1)
+        return 0;
+    if (ctx->pwdlen > ARGON2_MAX_PWD_LENGTH)
+        return 0;
+    if (ctx->salt == NULL && ctx->saltlen != 0)
+        return 0;
+    if (ctx->saltlen < ARGON2_MIN_SALT_LENGTH)
+        return 0;
+    if (ctx->saltlen > ARGON2_MAX_SALT_LENGTH)
+        return 0;
+    if (ctx->secret == NULL && ctx->secretlen != 0)
+        return 0;
+    if (ctx->secret != NULL && ARGON2_MIN_SECRET+1 > ctx->secretlen+1)
+        return 0;
+    if (ctx->secret != NULL && ARGON2_MAX_SECRET < ctx->secretlen)
+        return 0;
+    if (ctx->ad == NULL && ctx->adlen != 0)
+        return 0;
+    if (ctx->ad != NULL && ARGON2_MIN_AD_LENGTH+1 > ctx->adlen+1)
+        return 0;
+    if (ctx->ad != NULL && ARGON2_MAX_AD_LENGTH < ctx->adlen)
+        return 0;
+    if (ctx->m_cost < ARGON2_MIN_MEMORY)
+        return 0;
+    if (ctx->m_cost >= (uint64_t) ARGON2_MAX_MEMORY)
+        if (ctx->m_cost != (uint64_t) ARGON2_MAX_MEMORY)
+            return 0;
+    if (ctx->m_cost < 8 * ctx->lanes)
+        return 0;
+    if (ctx->t_cost < ARGON2_MIN_TIME)
+        return 0;
+    if (ctx->t_cost > ARGON2_MAX_TIME)
+        return 0;
+    if (ctx->lanes < ARGON2_MIN_LANES)
+        return 0;
+    if (ctx->lanes > ARGON2_MAX_LANES)
+        return 0;
+
+#if !defined(ARGON2_NO_THREADS) && defined(OPENSSL_THREADS)
+    if (ctx->threads < ARGON2_MIN_THREADS)
+        return 0;
+    if (ctx->threads > ARGON2_MAX_THREADS)
+        return 0;
+    if (ctx->threads > 1 && CRYPTO_THREAD_enabled(ctx->openssl_ctx) != 1)
+        return 0;
+    if (ctx->threads > 1 && ctx->threads >
+            crypto_thread_get_available_threads(ctx->openssl_ctx))
+        return 0;
+#else
+    if (ctx->threads != 1)
+        return 0;
+#endif
+
+    return 1;
+}
+
+static void initial_hash(uint8_t *blockhash, ARGON2_CTX *ctx)
+{
+    EVP_MD_CTX *mdctx;
+    uint8_t value[sizeof(uint32_t)];
+    unsigned int tmp;
+    uint32_t args[] = {
+        ctx->lanes, ctx->outlen, ctx->m_cost, ctx->t_cost, ctx->version,
+        (uint32_t) ctx->type, ctx->pwdlen
+    };
+
+    if (ctx == NULL || blockhash == NULL)
+        return;
+
+    mdctx = EVP_MD_CTX_create();
+    if (mdctx == NULL)
+        goto fail;
+
+    if (EVP_DigestInit_ex(mdctx, EVP_blake2b512(), NULL) != 1)
```

Comment:
This is not library context friendly.
blake2b512 should be fetched using the ctx->openssl_ctx

Reply:
Fixed, thanks!